pkg/test/framework/driver.go

```diff
@@ -12,153 +12,117 @@
 
-var scope = log.RegisterScope("testframework", "General scope for the test framework", 0)
+	// The driver is currently running tests.
+	running
+
+	// the driver has completed running tests.
+	completed
+)
 
 type driver struct {
 	lock sync.Mutex
 
-	args *Args
-	ctx  *internal.TestContext
+	settings *settings.Settings
+	ctx      *internal.TestContext
 
 	// The names of the tests that we've encountered so far.
 	testNames map[string]struct{}
 
-	running bool
+	state state
 }
 
-var _ Interface = &driver{}
-
 // New returns a new driver instance.
-func New() Interface {
+func newDriver() *driver {
 	return &driver{
 		testNames: make(map[string]struct{}),
+		state:     created,
 	}
 }
 
-// Initialize implements same-named Interface method.
-func (d *driver) Initialize(a *Args) error {
-	scope.Debugf("Enter: driver.Initialize (%s)", a.TestID)
+// Run implements same-named Driver method.
+func (d *driver) Run(testID string, m *testing.M) (int, error) {
 	d.lock.Lock()
-	defer d.lock.Unlock()
-
-	if d.ctx != nil {
-		return errors.New("test driver is already initialized")
-	}
-
-	// Make a copy of the args
-	args := &(*a)
-	if err := args.Validate(); err != nil {
-		return err
-	}
-
-	// Initialize the environment.
-	var env internal.Environment
-	switch a.Environment {
-	case EnvLocal:
-		env = local.NewEnvironment()
-	case EnvKube:
-		env = cluster.NewEnvironment()
-	default:
-		return fmt.Errorf("unrecognized environment: %s", a.Environment)
-	}
-
-	// Create the context, but do not attach it to the member fields before initializing.
-	ctx := internal.NewTestContext(
-		args.TestID,
-		generateRunID(args.TestID),
-		a.WorkDir,
-		a.Hub,
-		a.Tag,
-		a.KubeConfig,
-		env)
-
-	if err := env.Initialize(ctx); err != nil {
-		return err
+	// do not defer unlock. We will explicitly unlock before starting test runs.
+	if d.state != created {
+		d.lock.Unlock()
+		return -1, fmt.Errorf("driver.Run must be called only once")
 	}
+	d.state = running
 
-	d.ctx = ctx
-	d.args = args
-
-	return nil
-}
-
-// Run implements same-named Interface method.
-func (d *driver) Run() int {
-	scope.Debugf("Enter: driver.Run (%s)", d.ctx.TestID())
-	d.lock.Lock()
-
-	// If context is not set, then we're not initialized.
-	if d.ctx == nil {
+	s, err := settings.Construct(testID)
+	if err != nil {
 		d.lock.Unlock()
-		scope.Error("test driver is not initialized yet")
-		return -1
+		return -1, err
 	}
+	scope.Debugf("driver settings: %+v", s)
+	d.settings = s
 
-	if d.running {
+	if err = d.initialize(); err != nil {
 		d.lock.Unlock()
-		scope.Error("test driver is already running")
-		return -2
+		return -2, err
 	}
 
-	for _, dep := range d.args.SuiteDependencies {
+	for _, dep := range d.settings.SuiteDependencies {
 		if err := d.ctx.Tracker().Initialize(d.ctx, d.ctx.Environment(), dep); err != nil {
-			log.Errorf("Failed to initialize dependency '%s': %v", dep, err)
-			return -3
+			scope.Errorf("driver.Run: Dependency error '%s': %v", dep, err)
```

Comment:
do you need to unlock here?

I wonder if we should break this into a few functions, to avoid these sorts of errors:

```golang
func (d *driver) preRun() (int, error) {
  d.lock.Lock()
  defer d.lock.Unlock()

  if d.state != created {
    return -1, fmt.Errorf("driver.Run must be called only once")
  }
  d.state = running

  s, err := settings.Construct(testID)
  if err != nil {
    return -1, err
  }
  scope.Debugf("driver settings: %+v", s)
  d.settings = s

  if err = d.initialize(); err != nil {
    return -2, err
  }

  for _, dep := range d.settings.SuiteDependencies {
    if err := d.ctx.Tracker().Initialize(d.ctx, d.ctx.Environment(), dep); err != nil {
      scope.Errorf("driver.Run: Dependency error '%s': %v", dep, err)
      return -3, err
    }
  }
}

func (d *driver) postRun() {
  d.lock.Lock()
  defer d.lock.Unlock()

  d.state = completed

  if !d.settings.NoCleanup {
    d.ctx.Tracker().Cleanup()
  }
}

func (d *driver) Run(testID string, m *testing.M) (int, error) {
  rt, err := d.preRun(testID, m)
  if err != nil {
    return rt, err
  }

  // Call m.Run() while not holding the lock.
  lab.Infof(">>> Beginning test run for: '%s'", testID)
  rt = m.Run()
  lab.Infof("<<< Completing test run for: '%s'", testID)

  d.postRun()
  return rt, nil
}
```